telegram/_bot.py

```diff
@@ -1157,7 +1165,10 @@ async def send_audio(
 
         """
         thumbnail_or_thumb: FileInput = warn_about_thumb_return_thumbnail(
-            deprecated_arg=thumb, new_arg=thumbnail
+            deprecated_arg=thumb,
+            new_arg=thumbnail,
+            warn_callback=self._warn,
+            stacklevel=4,
```

Comment:
I just thought of one thing: could this indication of stacklevel be some sort of violation of Demeter law or a similar rule? It's like a class whose method is being called knows something about what is calling it and through how many "layers".

Reply:
mh, the only thing this method knows is how many layers *inside* of the `Bot` class there are between this line and the actual interface of the class. IMO that's no forbidden info …

Reply:
so a `Bot` being an attribute of `Application` doesn't affect `stacklevel`?

Reply:
No. we just want to know how many layers there are between `await bot.send_photo` and `warnings.warn`. In this case the stack is:

1. the call to `bot.send_photo`
2.  the `@_log` decorator
3. the `warn_about_thumb_return_thumbnail` function
4. the `warn_about_deprecated_arg_return_new_arg` function
5. * the `warn_callback` argument of that function, here `self._warn`
6. * the function `telegram._utils.warnings.warn`
7. the actual call to `warnings.warn`

To issue the warning form the `bot.send_photo` line, we need a stacklevel of 6 - that is 6 layers above `warnings.warn`. The two calls marked with * basically already have this "Demeter law" in mind - they add + 1 to the stacklevel so that you don't have to care about them from where you call them.
One could indeed argue that this should also be the case for the two `warn_about_*` functions - you're right here. If we do that, then we'd have to set a stacklevel of 2 in line 1171, which then only "knows about" the `@_log` decorator. That's really a `Bot`-class internal thingy though and whoever issues a warning from somewhere within `Bot` can see if they need to consider the decorator or not.

Reply:
Great, thanks for the explanation!